pybossa.js

```diff
@@ -166,8 +167,122 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+            //This flag is to determine when to show the passedQuizNotification
+            window.pybossa.passedQuizShowed = true;
+            return { msg: 'You must complete a quiz successfully before you can work on this job. You are in Quiz mode.',
+                     type: 'info' }
+        }
+    }
+
+    // This func determine if there are no more questions available for the quiz.
+    function _outOfGoldenTasks (userProgress, quiz, config){
+        if (quiz && config.enabled && userProgress.remaining_for_user === 0 && quiz.status === 'in_progress' && !window.pybossa.isGoldMode){
+            return { msg: 'We have run out of quiz questions for you. Please notify the project owner.',
+                     type: 'info' }
+        }
+    }
+
+    //This func is to determine if in Bulk gold mode are more task available to be converted into gold
+    function _outOfNonGoldTask (userProgress){
+        if (userProgress.available_gold_tasks === userProgress.total && window.pybossa.isBulk){
+            return { msg:'In gold mode, there are no tasks available.',
+                     type: 'warning' };
+        }
+    }
+
+    //This func determine if the quiz is failed and the user is not able to work on the project
+    function _failedQuiz (quiz, config){
+        if (quiz && config.enabled && quiz.status === 'failed'){
+            return { msg: 'Thank you for taking the quiz. You got ' + quiz.result.right + ' correct out of ' + quiz.config.questions +' tasks. ' +
+                            'You have been blocked from working on this job. The administrator of this job will contact you with next steps.',
+                     type: 'error' };
+        }
+    }
+
+    // This func determine if the user passed the quiz and use two flags to decide to show or not the notification since we only want to show it once
+    function _passedQuiz (quiz, config){
+        if (quiz && config.enabled && quiz.status === 'passed' && (window.pybossa.passedQuizShowed || window.pybossa.takingQuiz)){
+            window.pybossa.passedQuizShowed = false;
+            window.pybossa.takingQuiz = false;
+            return { msg: 'Thank you for taking the quiz. You got ' + quiz.result.right + ' correct out of ' + quiz.config.questions + ' tasks. ' +
+                           'You will now be able to work on this job.',
+                     type: 'warning' };
+        }
+    }
+
+    //This func determine if the project is completed.
+    function _projectCompleted (userProgress, quiz){
+        if (quiz && userProgress.remaining_for_user === 0 && quiz.status !== 'in_progress'){
+            return { msg: 'Congratulations, you have completed the job.',
+                     type: 'success' };
+        }
+    }
+
+    //This func determine if its in gold mode.
+    function _inGoldMode (userProgress){
+        if ((window.pybossa.isGoldMode && !(userProgress.available_gold_tasks === userProgress.remaining)) ||
```

Comment:
`available_gold_tasks` returns the number of unexpired gold tasks in the project without regard to whether the current user has answered them. `remaining` returns the number of tasks in the project that are still available for the current user to answer, but it doesn't consider expiration. So it seems that if the user answers some of the gold tasks and then goes into bulk gold mode then this code will stop them before they have converted all the tasks to gold. Meaning it will stop them when the number of gold tasks equals the number of tasks remaining for the user, not when the number of gold tasks equals the total number of tasks.

I don't know what the right way to do this is. Maybe we need to return `non_gold_tasks`  in `userProgress`.